src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,381 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Features.Consensus.CoinViews;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.ProvenBlockHeaders
+{
+    public class ProvenBlockHeaderStore : IProvenBlockHeaderStore, IDisposable
+    {
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Allows consumers to perform clean-up during a graceful shutdown.</summary>
+        private readonly INodeLifetime nodeLifetime;
+
+        /// <summary>Chain state holds various information related to the status of the chain and its validation.</summary>
+        private readonly IChainState chainState;
+
+        /// <summary>Thread safe class representing a chain of headers from genesis.</summary>
+        private readonly ConcurrentChain chain;
+
+        /// <summary>Database repository storing <see cref="ProvenBlockHeader"></see>s.</summary>
+        private readonly IProvenBlockHeaderRepository provenBlockHeaderRepository;
+
+        /// <summary>Lock object to protect access to <see cref="ProvenBlockHeader"/>.</summary>
+        private readonly AsyncLock lockobj;
+
+        /// <summary>Performance counter to measure performance of the save and query operation.</summary>
+        private readonly BackendPerformanceCounter performanceCounter;
+
+        private BackendPerformanceSnapshot latestPerformanceSnapShot;
+
+        /// <summary>Current block tip hash and height that the <see cref= "ProvenBlockHeader"/> belongs to.</summary>
+        public HashHeightPair CurrentTipHashHeight { get; private set; }
+
+        /// <summary>Pending - not yet saved to disk - block tip hash and height that the <see cref= "ProvenBlockHeader"/> belongs to.</summary>
+        private HashHeightPair pendingTipHashHeight;
+
+        /// <summary>The async loop we wait upon.</summary>
+        private IAsyncLoop asyncLoop;
+
+        /// <summary>Factory for creating background async loop tasks.</summary>
+        private readonly IAsyncLoopFactory asyncLoopFactory;
+
+        /// <summary>Maximum number of items to cache.</summary>
+        private const int CacheMaxItemsCount = 1_000;
```

Comment:
There should only be size in bytes limitation for cache. 

Count limitation of <2000 introduces a DoS vector since every time we will need to construct headers we will always go to the database.

Reply:
> There should only be size in bytes limitation for cache.

Bytes limitation also exists.  I needed some value to construct the LRU Cache as it needs to take in max items.

> Count limitation of <2000 introduces a DoS vector since every time we will need to construct headers we will always go to the database.

Shall I up the count to 2000 then ?

Reply:
Removed this now working on size rather then item count.